### ISSUE TITLE
grid_sample is not aligned

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#20785]
`F.interpolate` has been changed to use `align_corners=False` by default, which is also the pixel model used by opencv/PIL, etc.

The input to `grid_sample` is in range `[-1, 1]`, and therefore should ideally be scale-invariant.
However, `F.grid_sample` still uses the old pixel model: https://github.com/pytorch/pytorch/blob/65b00aa5972e23b2a70aa60dec5125671a3d7153/aten/src/ATen/native/cuda/GridSampler.cu#L161-L162, and as a result it is not properly aligned under the scaling of `F.interpolate(align_corners=False)`. 

Examples:

```python
import torch
from torch.nn import functional as F

def grid_sample_v2(input, grid):
    # grid: [-1, 1]
    N, C, H, W = input.shape
    gridx = grid[:, :, :, 0]
    gridy = grid[:, :, :, 1]
    gridx = ((gridx + 1) / 2 * W - 0.5) / (W - 1) * 2 - 1
    gridy = ((gridy + 1) / 2 * H - 0.5) / (H - 1) * 2 - 1
    newgrid = torch.stack([gridx, gridy], dim=-1)
    return F.grid_sample(input, newgrid)

N = 500
input = torch.rand(N, 1, 32, 74)
input2 = F.interpolate(input, scale_factor=5, align_corners=False, mode='bilinear')

grid = torch.rand(N, 18, 17, 2) * 1.8 - 0.9  # use coordinates in -0.9, 0.9 to avoid boundary effects

output = F.grid_sample(input, grid)
output2 = F.grid_sample(input2, grid)
diff = torch.abs(output2 - output)
print("DIFFv1:", diff.mean())

output = grid_sample_v2(input, grid)
output2 = grid_sample_v2(input2, grid)
diff = torch.abs(output2 - output)
print("DIFFv2:", diff.mean())
```
The above code, when run with `align_corners=True`, prints:
```
DIFFv1: tensor(0.0032)
DIFFv2: tensor(0.0675)
```
when run with `align_corners=False`, prints:
```
DIFFv1: tensor(0.0689)
DIFFv2: tensor(5.3741e-07)
```
In the example, I provide a new implementation of `grid_sample`, which leads to smaller reconstruction error when `align_corners=False`. Given that `align_corners=False` is now the default for `interpolate`, similar options should be built into `grid_sample` as well.

**Comment by bnehoran** [comment]
Yes. I agree that the convention for `F.grid_sample` should be updated to match that of `align_corners=False` in `F.interpolate`.
This would allow `F.grid_sample` to be agnostic to the size of the sampled image.

The way it is currently defined, -1 and 1 refer to the centers of the edge pixels, which is not a consistent location on images of different sizes. To make this size-agnostic, these should refer to the edges of the image themselves, which doesn't change upon upsampling/downsampling the image.

I have been working for almost a year now with a wrapped version of `F.grid_sample` which does just that, similar to what you use, since I have needed very precise behavior for my research. Mine is a bit more simplified, since note, for example, that in the wrapping function it is sufficient to multiply the grid by `size` and divide by `size - 1` (your conversion simplifies to this if you work it out analytically).
It seems appropriate, however, that this should be the default behavior.

If we decide to update the convention here, then we should also update the convention in `F.affine_grid` to match this as well.

One potential issue with updating the convention is that it would break any existing code that uses `F.grid_sample` on grids not generated from `F.affine_grid`. A possible way forward could be to do something similar to the way `F.upsample` and `F.interpolate` were changed, by adding a flag that reverts to the old behavior, so that users can keep the existing convention if they choose.

@SsnL What do you think?

**Comment by bnehoran** [comment]
In case it helps to see it visually, here is part of a presentation I gave last August that referenced this problem:
https://docs.google.com/presentation/d/1sjPK4DwIuI0HUHF4jvwlErxCaDVidsybt_uNdrB7aog/edit?usp=sharing

The important slides are this one, which shows the -1 and +1 reference moving around when upsampling/downsampling with the current `grid_sample` convention:
![image](https://user-images.githubusercontent.com/9757500/58150486-c5315900-7c34-11e9-9466-24f2bd431fa4.png)
and this one, which shows the size-agnostic solution
![image](https://user-images.githubusercontent.com/9757500/58150604-222d0f00-7c35-11e9-976c-ed452ab13cec.png)

**Comment by ezyang** [comment]
cc @fmassa

**Comment by fmassa** [comment]
I agree with this problem, and have reached the same conclusion in the past as well.
We could add another option to grid_sample, `align_corners`, like we did with `interpolate`, to handle both behaviors

**Comment by SsnL** [comment]
I agree with @fmassa that an option would be great. Whether we want to make it default is unclear, as the current implementation aligns with cudnn.

**Comment by bnehoran** [comment]
Unfortunately, it seems that the coordinate conventions specified in the original STN paper were a bit ambiguous, and so different deep learning frameworks ended up interpreting and implementing them slightly differently. And most ended up getting them not quite right.

@SsnL do you know where the conventions for the cudnn STN implementation are specified? I understand it's not open sourced.

If we were to make this the default, would it be too messy if in the cudnn version, it simply multiplied by the conversion factor before passing it on to the cudnn function?
Are you concerned more that this could cause a slight performance hit, or rather that the default would no longer match up with the result from using cudnn directly?

Either way, I would be willing to help make these changes (adding the `align_corners` option and functionality, etc.) if we decide that this is the way to go.

**Comment by fmassa** [comment]
@bnehoran we can dispatch to our own kernels if `align_corners=False`, and disable cudnn in this case for now. We have done that in the past when we added new options to `grid_sample`, like the padding mode and the nearest interpolation modes.

---

There is also another point that I'd like to make about `grid_sample`, which is that even after we make this change to make it aligned, it won't match 1:1 with `interpolate` in some cases.

Indeed, if we use `grid_sample` to downsample an image using bilinear interpolation, it will always take the 4 closest pixels that correspond to the neighbors in the image space. This means that for large downsampling factors, this will make the bilinear interpolation look almost like a nearest neighbor interpolation.

Here is where this is defined
https://github.com/pytorch/pytorch/blob/90182a7332997fb0edf666abc4b554b83a1670d1/aten/src/ATen/native/cuda/GridSampler.cu#L181-L186
We might potentially want to replace the `+1` with the corresponding image offset after a lookup in the coordinates of the next elements in the grid, to have smoother interpolation. But I haven't thought about it throughly, and we might be overfitting to a particular case. In particular, I don't know how what I proposed just above would behave with grids having holes.

I understand that this is completely unrelated to the original issue, but I raised it here for awareness. It might be better to open a new issue about this.

**Comment by soumith** [comment]
i think we should remove CuDNN bindings and do the right thing. Our own kernels aren't that much slower than CuDNN, and if we actually wanted to we can speed them up a bit more.

**Comment by bnehoran** [comment]
Is this a go-ahead to start working on it?
I am currently finishing some work on a reverse grid sampler, but I can probably start putting time into this sometime over the next few days.
I think I can figure out many of the places where changes need to be made, but I will probably need help with some parts, such as adding tests for the new functionality.

**Comment by bnehoran** [comment]
@fmassa I am not sure I understand how what you are suggesting would work for a more general grid.
First of all, if I understand what you are suggesting, you are proposing to allow `grid_sample` to also do the job of upsampling and/or downsampling tensors, right?
Just to avoid confusion, as you mention, this is different than the point made in this issue, which is more about using the functionalities in conjunction with one another (that is, using `grid_sample` on a tensor that has been upsampled or downsampled using `interpolate`), but incidentally, it seems that these changes should also cover the case of using `gird_sample` to upsample/downsample, for the most part.

I agree that calling `grid_sample` on a tensor using an identity grid of larger size than the tensor should have the same effect as upsampling it using `interpolate` with bi/tri/linear modes, and I think these changes should do that. Similarly, attempting to downsample a tensor by using `grid_sample` with an identity grid that is smaller than the tensor, would bi/tri/linearly interpolate between the nearest whole pixels (note: rather than average pooling over the nearby area), which I believe should also be equivalent to the bi/tri/linear modes of `interpolate`. I think this equivalence is what you meant to point out, right?
The `nearest` modes should similarly become equivalent.

There would just be no equivalent to the `area` mode of `interpolate`. However, it's not immediately clear to me how such a mode should really behave for `grid_sample`. If the identity grid is evenly spaced, then it is all well, since the rectangular areas that each grid pixel covers are well defined. However, as soon as the grid starts deviating significantly from the identity (for example, in an extreme case, suppose that multiple grid points land on the same location of the sampled tensor), how do you then define the "area" covered by each grid point?
Maybe there does exist some way to define it that is both natural and consistent, but at the very least, it seems non-trivial.

**Comment by fmassa** [comment]
@bnehoran I don't think there is anyone currently working on adding an `align_corners` option to `grid_sample`, so feel free to take it and let us know if you have questions.

About the next point that you mentioned (and that I originally raised in my last message), I'm also not 100% about how to define it properly as of now. Let's move the discussion to https://github.com/pytorch/pytorch/issues/21457 instead.

**Comment by markeroon** [comment]
@bnehoran Is that reverse grid sampler going to available anywhere? :)

**Comment by bnehoran** [comment]
> @bnehoran Is that reverse grid sampler going to available anywhere? :)

Yeah, I guess I could make it available when it's ready. It's part of a larger package with a variety of different functionality relating to displacement fields and mappings (including a workaround for the problem raised in this GitHub issue).

To be more precise, the "reverse grid sampler" I'm working on is a differentiable algorithm that produces a quite accurate inverse displacement field (equivalently, a grid), which can then be used with the grid sampler to retransform an image tensor back to its original coordinates (with several caveats, of course).
My use case, though, is to train neural networks that predict displacements in the Lagrangian direction, rather than the standard Eulerian direction, since this has advantages for certain applications.

@markeroon If you let me know what your use case is, I'll see if I could publish it in a way that would be useful to you.

**Comment by bnehoran** [comment]
I've opened a pull request at https://github.com/pytorch/pytorch/pull/23923 to add the `align_corners` option and functionality as discussed here.

@SsnL and/or @fmassa I would appreciate if you gave it a look!

**Comment by MingyangLei** [comment]
> In case it helps to see the issue visually, there is part of a presentation I gave last August that referenced this problem and its solution which could be useful:
> https://docs.google.com/presentation/d/1sjPK4DwIuI0HUHF4jvwlErxCaDVidsybt_uNdrB7aog/edit?usp=sharing
> 
> The important slides are this one, which shows the -1 and +1 reference moving around when upsampling/downsampling with the current `grid_sample` convention:
> ![image](https://user-images.githubusercontent.com/9757500/58150486-c5315900-7c34-11e9-9466-24f2bd431fa4.png)
> and this one, which shows the size-agnostic solution
> ![image](https://user-images.githubusercontent.com/9757500/58150604-222d0f00-7c35-11e9-976c-ed452ab13cec.png)
> 
> You can also look at the speaker notes if you want a more thorough description.

Has this F.grid_sample problem been fixed in the latest PyTorch? How can I keep it size-agnostic while applying F.grid_sample? Thanks!

**Comment by krikru** [comment]
> In case it helps to see the issue visually, there is part of a presentation I gave last August that referenced this problem and its solution which could be useful: https://docs.google.com/presentation/d/1sjPK4DwIuI0HUHF4jvwlErxCaDVidsybt_uNdrB7aog/edit?usp=sharing

Great illustrations bnehoran, thank you very much! Has the default behavior in PyTorch's [`grid_sample`](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html) function been changed now? It's a bit hard to tell what the default behavior is, since the documentation of the function head says `align_corners=None`, but the documentation for the parameter says `Default: False`. This inconsistency is slightly confusing, and one of these two is obviously wrong. If the default value is `None`, and `None` is interpreted as `False`, I think the documentation should say that instead (if only to make the documentation more consistent).

Anyway, interesting how TensorFlow chooses to do it. I think the coordinates their function takes are closer to what its internal implementation needs, and therefore is just *slightly* more computationally efficient than PyTorch's implementation (when `align_corners=True`, PyTorch needs to remove `1` from `IW` and `IH` when calculating the "normalized" indices, even though that' probably negligible, and when `align_corners=False`, although it's probably more intuitive and elegant to the user of the function, PyTorch needs to remove `0.5` from every index, which I guess is more taxing). So from that perspective, I think their implementation makes some sense at least.

However, what probably makes the most sense from a computational perspective would be to just let the user give coordinates that were already normalized, so `grid_sample` wouldn't have to do that at all. Why force the user to provide the function with coordinates that are in the range -1 to 1, if that just means it will have to do extra work and transform them internally to some other ranges anyway? If we assume the user only have the coordinates in that range, I guess that is fine, but if the user already has them in the ranges the function transforms them to internally we end up doing more computation than we would have to. I think there should be an option for the user to give coordinates that are already normalized, to get code that is computationally more efficient, if that's what the user wants.
